src/shared/Microsoft.Git.CredentialManager/Application.cs

```diff
@@ -147,63 +146,41 @@ Task IConfigurableComponent.ConfigureAsync(ConfigurationTarget target)
+                    ? GitConfigurationLevel.System
+                    : GitConfigurationLevel.Global;
 
-                        // Clear any existing app entries in the configuration
-                        config.UnsetAll(helperKey, Regex.Escape(appPath));
+            Context.Trace.WriteLine($"Configuring for config level '{configLevel}'.");
 
-                        // Add an empty value for `credential.helper`, which has the effect of clearing any helper value
-                        // from any lower-level Git configuration, then add a second value which is the actual executable path.
-                        config.ReplaceAll(helperKey, Constants.RegexPatterns.None, string.Empty);
-                        config.ReplaceAll(helperKey, Constants.RegexPatterns.None, appPath);
-                    }
-                    break;
+            IGitConfiguration config = Context.Git.GetConfiguration(configLevel);
 
-                case ConfigurationTarget.System:
-                    // For machine-wide configuration, we are looking for the following to be set in the system config:
-                    //
-                    // [credential]
-                    //     helper = {appPath}
-                    //
-                    config = Context.Git.GetConfiguration(GitConfigurationLevel.System);
-                    string currentValue = config.GetValue(helperKey);
-                    if (Context.FileSystem.IsSamePath(currentValue, appPath))
-                    {
-                        Context.Trace.WriteLine("Credential helper system configuration is already set correctly.");
-                    }
-                    else
-                    {
-                        Context.Trace.WriteLine("Updating Git credential helper system configuration...");
-                        config.SetValue(helperKey, appPath);
-                    }
+            // We are looking for the following to be set in the config:
+            //
+            // [credential]
+            //     ...                # any number of helper entries (possibly none)
+            //     helper =           # an empty value to reset/clear any previous entries (if applicable)
+            //     helper = {appPath} # the expected executable value & directly following the empty value
+            //     ...                # any number of helper entries (possibly none)
+            //
+            string[] currentValues = config.GetRegex(helperKey, Constants.RegexPatterns.Any).ToArray();
 
-                    break;
+            // Try to locate an existing app entry with a blank reset/clear entry immediately preceding
+            int appIndex = Array.FindIndex(currentValues, x => Context.FileSystem.IsSamePath(x, appPath));
+            if (appIndex > 0 && string.IsNullOrWhiteSpace(currentValues[appIndex - 1]))
+            {
+                Context.Trace.WriteLine("Credential helper configuration is already set correctly.");
```

Comment:
Should we also look for empty entries _after_ this (which essentially disable GCM Core)?

Reply:
True.. will also need to check no other empty entries follow `appIndex`.

Reply:
Done. If there are empty entries after the GCM entry, we will attempt to add GCM back 'in front' (with another "" value to make sure GCM is the only helper) again.